### PR TITLE
Issue 22

### PR DESCRIPTION
**Pull request by koheiw**
Generalizing the solution in #21 and enable `use_nan` for other methods. This change should only affect "cosine", "kullback, and "chisquared".

I also updated the [README](https://github.com/koheiw/proxyC/tree/issue-22) with budges, I was surprised to see that our package is downloaded 8K times every month.

### CODE REVIEW

File: src/linear.cpp [hunk 1]
```diff
@@ -74,22 +75,18 @@ struct linearWorker : public Worker {
                 v1 = rowvec(trans(mt1t * mt2.col(i)));
                 v2 = center1 * center2[i] * ncol;
                 simils = to_vector(((v1 - v2) / ncol) / (square1 * square2[i]));
-                for (std::size_t j = 0; j < square1.size(); j++) {
-                    if (square1[j] == 0.0 || square2[i] == 0.0)
-                        simils[j] = use_nan ? std::numeric_limits<double>::quiet_NaN() : 0.0;
```
Comment (rcannood) [inline src/linear.cpp:79]: If you remove this, will you not get the same issue as in #20? Example:

``` r
mat1 <- Matrix::Matrix(1:4, nrow = 1, sparse = TRUE)
mat2 <- Matrix::Matrix(rep(0.1111, 4), nrow = 1, sparse = TRUE)
mat3 <- Matrix::Matrix(rep(1, 4), nrow = 1, sparse = TRUE)
simil(mat1, mat2, method = "correlation")
# [1,] Inf
simil(mat1, mat2, method = "correlation", diag = TRUE)
# [1,] Inf
simil(mat1, mat3, method = "correlation")
# [1,] .
simil(mat1, mat3, method = "correlation", diag = TRUE)
# [1,] 0

File: src/linear.cpp [hunk 2]
```diff
@@ -107,8 +104,6 @@ S4 cpp_linear(arma::sp_mat& mt1,
 
     if (mt1.n_rows != mt2.n_rows)
         throw std::range_error("Invalid matrix objects");
-    if (method != 2) // correlation only
-        use_nan = false;
```
Comment (rcannood) [inline src/linear.cpp:111]: See comment above.

File: src/pair.cpp [hunk 3]
```diff
@@ -9,23 +9,23 @@ double simil_cosine(colvec& col_i, colvec& col_j) {
     return accu(sum(col_i % col_j) / sqrt(sum(square(col_i)) * sum(square(col_j))));
 }
 
-double simil_correlation(colvec& col_i, colvec& col_j, bool use_nan) {
+double simil_correlation(colvec& col_i, colvec& col_j) {
     double sd_i = stddev(col_i, 1);
     double sd_j = stddev(col_j, 1);
-    if (sd_i == 0.0 || sd_j == 0.0)
-        return use_nan ? std::numeric_limits<double>::quiet_NaN() : 0.0;
```
Comment (rcannood) [inline src/pair.cpp:16]: If you remove this, will you not get the same issue as in #20? Example:

``` r
mat1 <- Matrix::Matrix(1:4, nrow = 1, sparse = TRUE)
mat2 <- Matrix::Matrix(rep(0.1111, 4), nrow = 1, sparse = TRUE)
mat3 <- Matrix::Matrix(rep(1, 4), nrow = 1, sparse = TRUE)
simil(mat1, mat2, method = "correlation")
# [1,] Inf
simil(mat1, mat2, method = "correlation", diag = TRUE)
# [1,] Inf
simil(mat1, mat3, method = "correlation")
# [1,] .
simil(mat1, mat3, method = "correlation", diag = TRUE)
# [1,] 0

File: src/pair.cpp [hunk 4]
```diff
@@ -220,8 +220,6 @@ S4 cpp_pair(arma::sp_mat& mt1,
 
     if (mt1.n_rows != mt2.n_rows)
         throw std::range_error("Invalid matrix objects");
-    if (method != 2) // correlation only
-        use_nan = false;
```
Comment (rcannood) [inline src/pair.cpp:224]: See comment above.

File: src/linear.cpp (hunk 1 above)
Comment (koheiw) [reply src/linear.cpp:79]: Thanks. I shouldn't have removed `replace_inf()` (f93fde0). I don't know why the tests did not detect this...

**Comment by koheiw** [comment]
@rcannood I noticed that the test matrix (`mat_test`) did not have rows/columns that we want to test the functions with. I addressed the issue.

I restored `replace_inf()` to return NaN instead of Inf. codedcov complains it is not tested, but src/proxyc.h#L86 is executed in [one of the tests](https://github.com/koheiw/proxyC/blob/0af460709d515b3067aa20ecd95aa8958c243926/tests/testthat/test-simil.R#L117).

**Review by rcannood — APPROVED** [review]

**Comment by rcannood** [comment]
My apologies for the last few commits ;) Everything looks good to me!

**Comment by koheiw** [comment]
@rcannood thanks for reviewing.

**Pull request merged**